### PR TITLE
core: pager: fix increment calculation in tee_pager_set_um_region_attr()

### PR DESCRIPTION
**Pull request by jenswi-linaro**
With the introduction of the commit referred below may a paged region
span multiple translation tables, but tee_pager_set_um_region_attr()
wasn't updated to handle this. So fix this by by accurately calculate
the increment in the main loop of that function.

Fixes: 4a3f6ad054d4 ("core: pager: let struct tee_pager_area span multiple translation tables")
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome@forissier.org> (HiKey)`

Tested as described in https://github.com/OP-TEE/optee_os/issues/4624#issue-894372838.

**Comment by jenswi-linaro** [comment]
Hopefully fixed the travis complaint. Tag applied.

**Comment by etienne-lms** [comment]
I tested, but got a data abort when loading StMM instance. I took a bit of time to cross-check, but that's it.
```
$ ./scripts/symbolize.py -d out/arm/core 
optee optee: OP-TEE: revision 3.13 (0914298b)
D/TC:0   tee_entry_exchange_capabilities:102 Dynamic shared memory is enabled
D/TC:0 0 core_mmu_entry_to_finer_grained:753 xlat tables used 4 / 5
D/TC:? 0 load_stmm:285 stmm load address 0x40002000
E/TC:0 0 
E/TC:0 0 Core data-abort at address 0x40414000 (translation fault)
E/TC:0 0  fsr 0x00000a07  ttbr0 0x2ffd6000  ttbr1 0x00000000  cidr 0x0
E/TC:0 0  cpu #0          cpsr 0x80000133
E/TC:0 0  r0 0x00780107      r4 0x00000000    r8 0x40002000   r12 0x00000000
E/TC:0 0  r1 0x40002000      r5 0x2ffe53a0    r9 0x00413000    sp 0x30074e78
E/TC:0 0  r2 0x40415000      r6 0x40414000   r10 0x40282000    lr 0x2ffc7443
E/TC:0 0  r3 0x40414000      r7 0x2ffe54ec   r11 0x2ffe5580    pc 0x30046310
E/TC:0 0 TEE load address @ 0x2ffc0000
E/TC:0 0 Call stack:
E/TC:0 0  0x30046310 load_stmm at core/arch/arm/kernel/stmm_sp.c:288
E/TC:0 0  0x3004dea3 tee_ta_init_session at core/kernel/tee_ta_manager.c:675
E/TC:0 0  0x30052659 entry_open_session at core/tee/entry_std.c:372
E/TC:0 0  0x30045953 std_smc_entry at core/arch/arm/kernel/thread_optee_smc.c:194
E/TC:0 0  0x2ffc16a8 thread_std_smc_entry at core/arch/arm/kernel/thread_optee_smc_a32.S:164
E/TC:0 0 Panic 'unhandled pageable abort' at ?:0 
E/TC:0 0 TEE load address @ 0x2ffc0000
E/TC:0 0 Call stack:
E/TC:0 0  0x2ffc1ea1 print_kernel_stack at core/arch/arm/kernel/unwind_arm32.c:109
E/TC:0 0  0x2ffc5ab9 __do_panic at core/kernel/panic.c:31
E/TC:0 0  0x2ffc1a7f get_fault_type at core/arch/arm/kernel/abort.c:469
E/TC:0 0  0x2ffc09e4 __thread_abort_common at core/arch/arm/kernel/thread_a32.S:724
```

**Comment by etienne-lms** [comment]
sorry, i bit busy today. i'll try to test a bit by this evening...

**Comment by jforissier** [comment]
@etienne-lms 

> I tested, but got a data abort when loading StMM instance.

Is it a regression introduced (or revealed) by this PR? (in other words, is `FETCH_HEAD^` OK?)

**Comment by jenswi-linaro** [comment]
I've added a fix for the problem described by @etienne-lms. I reproduced the problem on qemu_v8 with pager and stmm.
However, after this fix there's another problem surfacing that I'm currently investigating.

**Comment by jenswi-linaro** [comment]
Now that I look at the next problem it seems it might be caused by this last fix. Stay tuned.

**Comment by jenswi-linaro** [comment]
Another fix. StMM still fails to execute correctly.

**Comment by etienne-lms** [comment]
fyi: checked-out 4ab7c97e. StMM is successfully loaded but aborts later (user mode translation fault):
```
E/TC:? 0 User mode data-abort at address 0x40413ff4 (translation fault)
E/TC:? 0  fsr 0x00000a07  ttbr0 0x2ffd2000  ttbr1 0x00000000  cidr 0x0
E/TC:? 0  cpu #0          cpsr 0x00000130
E/TC:? 0  r0 0x40414000      r4 0x00000000    r8 0x00000000   r12 0x00000000
E/TC:? 0  r1 0x00000088      r5 0x00000000    r9 0x00000000    sp 0x40414000
E/TC:? 0  r2 0x00000000      r6 0x00000000   r10 0x00000000    lr 0x40002004
E/TC:? 0  r3 0x00000000      r7 0x00000000   r11 0x00000000    pc 0x40006798
E/TC:? 0  region  0: va 0x40000000 pa 0x2ffc0000 size 0x001000 flags ---R-X
E/TC:? 0  region  1: va 0x40002000 pa 0x00000000 size 0x280000 flags r-xR--
E/TC:? 0  region  2: va 0x40282000 pa 0x00000000 size 0x193000 flags rw-RW-
E/TC:? 0  region  3: va 0x40415000 pa 0x00000000 size 0x001000 flags rw-RW-
D/TC:? 0 stmm_enter_user_mode:129 stmm panicked with code 0xdeadbeef
D/TC:? 0 tee_ta_open_session:718 init session failed 0xffff3024
Unable to open OP-TEE session (err=-5)
mm_communicate failed!
Error: Cannot initialize UEFI sub-system, r = 3
STM32MP>

**Comment by jenswi-linaro** [comment]
This, I hope, is the last fix needed.

File: core/arch/arm/mm/tee_pager.c [hunk 1]
```diff
@@ -910,11 +911,20 @@ static struct pgt **alloc_merged_pgt_array(struct vm_paged_region *a,
 {
 	size_t a_next_pgt_count = get_pgt_count(a_next->base, a_next->size);
 	size_t a_pgt_count = get_pgt_count(a->base, a->size);
+	size_t pgt_count = get_pgt_count(a->base, a->size + a_next->size);
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c:914]: `_maybe_unused` for `pgt_count`

**Review by etienne-lms — COMMENTED** [review]
Last fix brought StMM back to live.
I've tested few configs and pager pool sizes. No issue found yet.

File: core/arch/arm/mm/tee_pager.c (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/mm/tee_pager.c:914]: Oh, I see that I forgot to replace one occurence of `get_pgt_count(a->base, a->size + a_next->size)`.

**Comment by jforissier** [comment]
@jbech-linaro IBART needs `sudo apt install libmpc-dev` too.

**Comment by jbech-linaro** [comment]
> @jbech-linaro IBART needs `sudo apt install libmpc-dev` too.

Done!

**Comment by etienne-lms** [comment]
for info: successfully tested in 32bit pager (StMM, regression and gp).
`Tested-by: Etienne Carriere <etienne.carriere@linaro.org> (stm32mp1 w/ gp, stmm)`
I would like spend time for reviewing the changes ASAIC.

**Comment by jenswi-linaro** [comment]
Squashed and tags applied. Note that these are just tested-by tags.

**Comment by jenswi-linaro** [comment]
Rebased on master without conflicts.

**Comment by jenswi-linaro** [comment]
What are we going to do about this PR? It's important fixes for problems likely to be seen when large TAs are used.

**Comment by etienne-lms** [comment]
sure, this is needed. i'll take some time to go through.
sorry for the delay.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"core: pager: fix increment calculation in tee_pager_set_um_region_attr()"
"core: pager: fix split_region() pgt assignment"

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"core: pager fix alloc_merged_pgt_array()"
"core: pager: fix split_region() to maintain order"

**Comment by jenswi-linaro** [comment]
Tags applied

**Comment by jforissier** [comment]
@etienne-lms thanks for reviewing, these fixes are badly needed in master :)

**Comment by etienne-lms** [comment]
thanks for your patience...

**Pull request merged**